File: Java/src/com/leetcode/Easy/PalindromeLinkedList.java
package com.leetcode.Easy;

import java.util.Stack;

// 234. Palindrome Linked List
public class PalindromeLinkedList {
    ListNode head;

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int value) {
            val = value;
            next = null;
        }
    }

    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null)
            return true;

        Stack<Integer> stack = new Stack<>();
        ListNode curr = head;
        while (curr != null) {
            stack.push(curr.val);
            curr = curr.next;
        }

        curr = head;
        while (curr != null) {
            int val = stack.pop();
            if (curr.val != val)
                return false;

            curr = curr.next;
        }
        return true;
    }
}

File: JavaScript/Medium/PalindromeLinkedList.js
/**
 * 234. Palindrome Linked List
 * @param {ListNode} head
 * @return {boolean}
 */
const isPalindrome = function (head) {
  if (head === null || head.next === null)
    return true;


  let slow = head, fast = head;
  while (fast !== null && fast.next !== null) {
    slow = slow.next;
    fast = fast.next.next;
  }

  let headSecondHalf = reverse(slow);
  let copyHeadSecondHalf = headSecondHalf;

  while (head !== null && headSecondHalf !== null) {
    if (head.val !== headSecondHalf.val) {
      return false;
    }
    head = head.next;
    headSecondHalf = headSecondHalf.next;
  }

  reverse(copyHeadSecondHalf);
  return true;
};

const reverse = function (head) {
  let prev = null;
  while (head !== null) {
    let next = head.next;
    head.next = prev;
    prev = head;
    head = next;
  }
  return prev;
};